app/components/Button/Redesign/Button.jsx

```diff
@@ -0,0 +1,62 @@
+// @flow
+
+import React, { Component } from 'react'
+import { omit } from 'lodash'
+
+import styles from './Button.scss'
+import Login from '../../../images/icons/Login.svg'
+import Plus from '../../../images/icons/Plus.svg'
+import Wallet from '../../../images/icons/Wallet.svg'
+
+const iconMap = name => {
+  if (!name) {
+    return console.warn('iconMap in Button.jsx invoked without a valid icon!')
+  }
+
+  const icons = {
+    login: Login,
+    plus: Plus,
+    wallet: Wallet
+  }
+  return icons[name]
+}
+
+type Props = {
+  icon: ?string,
+  text: string,
+  dark: ?boolean,
+  light: ?boolean
+}
+
+class Button extends Component<Props> {
+  render = () => {
+    const { dark, light, text, icon } = this.props
+    const passDownProps = omit(this.props, 'dark', 'light', 'icon')
+    const Icon = iconMap(icon)
```

Comment:
By hard coding the icon map in here, we're limiting the capabilities of the button.  Something I've started doing in components that may or may not want something rendered in a certain way (e.g.: a header or footer), is to pass in a special render function prop for that (e.g.: `renderHeader` or `renderFooter`).  That way, whatever is using the prop can pass in exactly what it needs using any style it wants, and there's no further need to update the other component to support new rendering functionality.

For this component, that might look something like this:

```js
type Props = {
  renderIcon: ?Function
}

render = () => {
  return (
    <button className={classNames(styles.button, buttonStyle)}>
      {this.renderIcon()}
      <span>{text}</span>
  )
}

renderIcon = {
  if (!this.props.renderIcon) {
    return null
  }

  return (
    <span className={classNames(styles.icon, iconStyle)}>
      {this.props.renderIcon()}
    </span>
  )
}
```

Reply:
I like this idea much better will refactor thanks for the suggestion

Reply:
Yeah, for sure!  Then you can just call it with something like this:

```js
<Button renderIcon={<Icon ... />}>Blah</Button>
```